cabal-install/src/Distribution/Client/CmdConfigure.hs

```diff
@@ -140,9 +149,101 @@ configureAction flags@NixStyleFlags {..} _extraArgs globalFlags = do
+     ++ "file (or '$project_file.local', if '--project-file' is specified) "
+     ++ "which extends the configuration from the 'cabal.project' file "
+     ++ "(if any).\n\n"
+
+     ++ "The v2-reconfigure command also checks that the project configuration "
+     ++ "will work. In particular it checks that there is a consistent set of "
+     ++ "dependencies for the project as a whole.\n\n"
+
+     ++ "It is never necessary to use the 'v2-reconfigure' command. It is "
+     ++ "merely a convenience in cases where you do not want to specify flags "
+     ++ "to 'v2-build' (and other commands) every time and yet do not want "
+     ++ "to alter the 'cabal.project' persistently.",
+  commandNotes        = Just $ \pname ->
+        "Examples:\n"
+     ++ "  " ++ pname ++ " v2-reconfigure --with-compiler ghc-7.10.3\n"
+     ++ "    Adjust the project configuration to use the given compiler\n"
+     ++ "    program and check the resulting configuration works.\n"
+     ++ "  " ++ pname ++ " v2-reconfigure\n"
+     ++ "    Reset the local configuration to empty and check the overall\n"
+     ++ "    project configuration works.\n"
+
+  , commandDefaultFlags = defaultNixStyleFlags ()
+  , commandOptions      = filter (\o -> optionName o /= "ignore-project")
+                        . nixStyleOptions (const [])
+  }
+
+reconfigureAction :: NixStyleFlags () -> [String] -> GlobalFlags -> IO ()
+reconfigureAction flags extraArgs globalFlags = do
+  reconf <- reconfigureAction' flags extraArgs globalFlags
+  let go (baseCtx, projConfig) = 
+        writeProjectLocalExtraConfig (distDirLayout baseCtx) projConfig
+
+  either go go reconf
+
+reconfigureAction' 
+    :: NixStyleFlags () 
+    -> [String] 
+    -> GlobalFlags 
+    -> IO (Either (ProjectBaseContext, ProjectConfig) (ProjectBaseContext, ProjectConfig))
+reconfigureAction' flags@NixStyleFlags {..} extraArgs globalFlags = do
+  baseCtx <- establishProjectBaseContext v cliConfig OtherCommand
+
+  let localFile = distProjectFile (distDirLayout baseCtx) "local"
+
+  exists <- doesFileExist localFile
+  if not exists
+    then do
+      notice v $
+        quote (takeFileName localFile)
+        <> " doesn't exist, calling configure instead."
+
+      -- @configureAction'@ is called, as opposed to @configureAction@, to avoid
+      -- calling on the writing function for configure, thusly maintaining the 
+      -- separation on the behaviour that modify files
+      Left <$> configureAction' flags extraArgs globalFlags
+  
+    else do
+      conf <- runRebuild (distProjectRootDirectory . distDirLayout $ baseCtx) $
+        readProjectLocalExtraConfig v (distDirLayout baseCtx)
+
+      buildCtx <- runProjectPreBuildPhase v baseCtx $ \plan -> 
+        return (plan, Map.empty)
+
+      printPlan v 
+        (baseCtx {buildSettings = (buildSettings baseCtx) {buildSettingDryRun = True}})
+        buildCtx
+
+      return $ Right (baseCtx, configurationMerge conf cliConfig)
```

Comment:
It doesn't seem useful here to lift these into an `Either`. You just end up calling `either go go` on them later, which implies the branching is unnecessary

Reply:
Ya, I actually don't like that either, no pun intended. The thing is, I used it that way because in a case within the `reconfigureAction'`, I'm calling `configureAction'`, that way I can differ one thing from the other when actually doing the tests.

Reply:
Coverage gives you that data. It tells you what branch you're going down and if it's covered by a particular test. I don't think you should embed branch coverage tests into the user-facing source code

Reply:
It's not a matter of what is being covered, it's about being able to tell when it was reconfigure or configure within reconfigure that was called. If those two are perceived as equals, then I do agree with you, otherwise I find it harmless to `Either`ise that, since it goes beyond the testing.

Reply:
However, I can see the wisdom in dismissing that as an `Either`, after all it does add another layer into the fold. If that does pose to be more complicated than it's worth down the road, I won't oppose its refactor.